packages/frontend-app-api/src/routing/resolveRouteBindings.ts

```diff
@@ -0,0 +1,145 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  RouteRef,
+  SubRouteRef,
+  ExternalRouteRef,
+} from '@backstage/frontend-plugin-api';
+import { RouteRefsById } from './collectRouteIds';
+import { Config } from '@backstage/config';
+
+/**
+ * Extracts a union of the keys in a map whose value extends the given type
+ *
+ * @ignore
+ */
+type KeysWithType<Obj extends { [key in string]: any }, Type> = {
+  [key in keyof Obj]: Obj[key] extends Type ? key : never;
+}[keyof Obj];
+
+/**
+ * Takes a map Map required values and makes all keys matching Keys optional
+ *
+ * @ignore
+ */
+type PartialKeys<
+  Map extends { [name in string]: any },
+  Keys extends keyof Map,
+> = Partial<Pick<Map, Keys>> & Required<Omit<Map, Keys>>;
+
+/**
+ * Creates a map of target routes with matching parameters based on a map of external routes.
+ *
+ * @ignore
+ */
+type TargetRouteMap<
+  ExternalRoutes extends { [name: string]: ExternalRouteRef },
+> = {
+  [name in keyof ExternalRoutes]: ExternalRoutes[name] extends ExternalRouteRef<
+    infer Params,
+    any
+  >
+    ? RouteRef<Params> | SubRouteRef<Params>
+    : never;
+};
+
+/**
+ * A function that can bind from external routes of a given plugin, to concrete
+ * routes of other plugins. See {@link createApp}.
+ *
+ * @public
+ */
+export type AppRouteBinder = <
+  TExternalRoutes extends { [name: string]: ExternalRouteRef },
+>(
+  externalRoutes: TExternalRoutes,
+  targetRoutes: PartialKeys<
+    TargetRouteMap<TExternalRoutes>,
+    KeysWithType<TExternalRoutes, ExternalRouteRef<any, true>>
+  >,
+) => void;
+
+/** @internal */
+export function resolveRouteBindings(
+  bindRoutes: ((context: { bind: AppRouteBinder }) => void) | undefined,
+  config: Config,
+  routesById: RouteRefsById,
+): Map<ExternalRouteRef, RouteRef | SubRouteRef> {
+  const result = new Map<ExternalRouteRef, RouteRef | SubRouteRef>();
+
+  if (bindRoutes) {
+    const bind: AppRouteBinder = (
+      externalRoutes,
+      targetRoutes: { [name: string]: RouteRef | SubRouteRef },
+    ) => {
+      for (const [key, value] of Object.entries(targetRoutes)) {
+        const externalRoute = externalRoutes[key];
+        if (!externalRoute) {
+          throw new Error(`Key ${key} is not an existing external route`);
+        }
+        if (!value && !externalRoute.optional) {
+          throw new Error(
+            `External route ${key} is required but was undefined`,
+          );
+        }
+        if (value) {
+          result.set(externalRoute, value);
+        }
+      }
+    };
+    bindRoutes({ bind });
+  }
+
+  const bindingsConfig = config.getOptionalConfig('app.routes.bindings');
+  if (!bindingsConfig) {
+    return result;
+  }
+
+  const bindings = bindingsConfig.get();
+  if (bindings === null || typeof bindings !== 'object') {
```

Comment:
`getOptionalConfig` already ensured this, should be ok to put a type arg on `get` and be done

Reply:
Ah yep, realized in tests but forgot to clean this bit up x)